inginious/frontend/plugins/upcoming_tasks/__init__.py

```diff
@@ -0,0 +1,154 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of INGInious. See the LICENSE and the COPYRIGHTS files for
+# more information about the licensing of this file.
+#
+
+""" A plugin that allow students to see their futur work in a single place for all their courses """
+import os
+from collections import OrderedDict
+from datetime import datetime, timedelta
+import flask
+
+
+from flask import send_from_directory
+from inginious.frontend.pages.utils import INGIniousPage, INGIniousAuthPage
+
+PATH_TO_PLUGIN = os.path.abspath(os.path.dirname(__file__))
+
+def menu(template_helper):
+    """ Displays the link to the board on the main page, if the plugin is activated """
+    return template_helper.render("main_menu.html", template_folder=PATH_TO_PLUGIN + '/templates/')
+
+class StaticMockPage(INGIniousPage):
+    """ MockPage based on auto-evaluation plugin structure """
+
+    def GET(self, path):
+        return send_from_directory(os.path.join(PATH_TO_PLUGIN, "static"), path)
+
+    def POST(self, path):
+        return self.GET(path)
+
+
+class UpComingTasksBoard(INGIniousAuthPage):
+
+    def GET_AUTH(self):
+        """ Called when reaching the page """
+        time_planner = "unlimited"
+        return self.page(time_planner)
+
+    def POST_AUTH(self):
+        """ Called when modifying time planner """
+        user_input = flask.request.form
+        time_planner = "unlimited"
+        if "time_planner" in user_input:
+            time_planner = user_input.get("time_planner")
+        return self.page(time_planner)
+
+    def time_planner_conversion(self, string_time_planner):
+        """ Used to convert the time_planner options into int value """
+        if string_time_planner == "unlimited":
+            return 100000
+        return int(string_time_planner)
+
+    def page(self, time_planner):
+        """ General main method called for GET and POST """
+        username = self.user_manager.session_username()
+        all_courses = self.course_factory.get_all_courses()
+        time_planner = self.time_planner_conversion(time_planner)
+
+        # Get the courses id
+        open_courses = {courseid: course for courseid, course in all_courses.items()
+                        if self.user_manager.course_is_open_to_user(course, username, False) and
+                        self.user_manager.course_is_user_registered(course, username)}
+        open_courses = OrderedDict(sorted(iter(open_courses.items()), key=lambda x: x[1].get_name(self.user_manager.session_language())))
+
+        courses = {self.course_factory.get_course(courseid): course for courseid, course in all_courses.items()
```

Comment:
Here you're using the Course object as dictionary key. I'd avoid that. In worst case it computes the hash based on a reference that can change, in best case it computes it somehow magically based on the complex data structure. 

I'm not even sure you need a dictionary because you don't seem to use key based setters and getters anywhere in the code.

Moreover, this line looks very similar to the ``open_courses`` one.

Reply:
Indeed, it was actually possible to manage everything only with open_courses without requiring a second dictonnary + it was, as you mentioned, using course as key instead of courseid.

The only little limitation I had with a single dictionnary is that I can not iterate on it with a for, while deleting items from it in the loop itself but I simply deleted items after the iterations and it's fine now.